lib/posix/pthread.c

```diff
@@ -254,34 +262,43 @@ int pthread_setcancelstate(int state, int *oldstate)
  */
 int pthread_cancel(pthread_t pthread)
 {
+	int rv = 0;
+	k_spinlock_key_t key;
+	bool should_abort = false;
 	struct posix_thread *thread = to_posix_thread(pthread);
-	int cancel_state;
-	k_spinlock_key_t cancel_key;
 
-	if ((thread == NULL) || (thread->state == PTHREAD_TERMINATED)) {
+	if (thread == NULL) {
```

Comment:
yeah, there's no protection here for using a thread index which is stale; checking only for one possible stale state (TERMINATED) provides a false sense of security. Makes me wonder if there should be some kind of validating mode for the API where a generation number is encoded in the upper bits of the thread ID that are validated against a current generation number stored in the thread?

Reply:
I like the thought and it would at least replicate some of the other validation schemes in e.g. `pthread_mutex.c`.

To be honest, I think I should be able to transition pthreads to use `<zephyr/sys/bitarray.h>` as well. 

Let me give it another spin.

Also, thanks @andyross and @npitre for validating the issue on aarch64. Will likely use an `arch_exclude` there if I am able to.